fms/utils/serialization.py

```diff
@@ -56,6 +66,97 @@ def list_sources(architecture: str):
+    weights (version 0.0.5)
+
+    Args:
+        name: str
+            current name to convert
+        orig_sd: Mapping
+            a mapping from a name to a param, in most cases a grouping of all names to params in a module.
+
+    Returns:
+    Optional[Tuple[str, torch.Tensor]]
+        if query/key/value all exist in the given state dict, a tuple of the new fused name as well as weights will be
+        returned from this function
+        otherwise this function will return None signifying that no preprocessing needed to be done for this name param
+    """
+    # if we find query/key/value, this means the weights are unfused and therefore need to be fused
+    if "attn.query" in name or "attn.key" in name or "attn.value" in name:
+        weight_type = name.split(".")[-1]
+
+        unfused_weights = [
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.query.{weight_type}",
+                name,
+            ),
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.key.{weight_type}",
+                name,
+            ),
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.value.{weight_type}",
+                name,
+            ),
+        ]
+
+        result = (
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.in_proj.qkv_fused.{weight_type}",
+                name,
+            ),
+            torch.cat([orig_sd[w] for w in unfused_weights], dim=0),
+        )
+
+        return result
+    return None
+
+
+def simple_mapping_adapter(
+    mappers: List[Callable[[str, Mapping], Optional[Tuple[str, torch.Tensor]]]],
+) -> Callable[[Mapping], Mapping]:
+    """
+    Create a simple adapter mapping using a mapper function. This function will iterate through the items in a
+    state_dict and execute the mapper on each key. This function could be used for versioning, where each version
+    requires a new mapping from the previous version.
+
+    Parameters
+    ----------
+    mappers: List[Callable[[str, Mapping], Optional[Tuple[str, torch.Tensor]]]]
+        a list of function which given a param key and the original state dict, will optionally return the new name and param
+        to be set in the new state dict. If None is returned, no change will be made in the current iteration
+
+    Returns
+    -------
+    Callable[[Mapping], Mapping]
+        the adapter function which gets the new state dict with the items re-mapped
+    """
+
+    def mapping_adapter_func(orig_sd):
+        new_sd = {}
+        for name, param in orig_sd.items():
+            for mapper in mappers:
+                mapper_out = mapper(name, orig_sd)
+
+                if mapper_out is not None:
+                    name, param = mapper_out
+                    break
```

Comment:
won't this prevent multiple consecutive adapters from being applied? is this supposed to only apply the first adapter in the list?

Reply:
You could theoretically nest each one of these to have them run (though that was not really the intention), will look into this as I know we added support for nested adapters in previous PRs.

Reply:
Looks like the intention was to simulate doing if / elif conditions inside the loop (changing a single parameter) which could happen pretty frequently, so this is not preventing multiple adapters. If you wanted to nest adapters, you could also create a singleton list with this and nest 2 of these (though that would be less efficient since it will go through the params multiple times - one for each adapter) -- that is what the current adapter would look like if we nested it.

Reply:
Looking more closely at the doc above for `mappers`, that seems like a really confusing API.

Is there any way for this to remove the original values from the sd? I think if we're converting k/q/v to kqv, once we have kqv we don't want to retain the original k/q/v tensors.

Reply:
I'll look into how to clean up the API. Also, I believe we should be able to remove weights that we are done with, I'll add that.